eluthia/zipapp_script.py

```diff
@@ -0,0 +1,187 @@
+            if len(os.listdir(history_folder)) > preinst_file_count:
+                return True
+        return False
+
+    missing_packages = [package for package in packages if get_version(package) == None] # packages in new_packages that aren't installed
+
+    if missing_packages:
+        raise InstallationError(f'Packages were not installed: {missing_packages}')
+
+    if not verify_badtrack(preinst_file_count, history_folder, timeout, check_frequency):
+        raise InstallationError('No files were generated by badtrack within 1 minute')
+
+    return
+
+
+def rollback(new_packages, old_packages):
+    """
+    Rolls back the installation of packages.
+    Input:
+        list new_packages - installed packages to be replaced
+        list old_packages - paths to old debian files to replace the installed versions
+    """
+    subprocess.run(['dpkg', '-r'] + list(new_packages))
+    subprocess.run(['dpkg', '--skip-same-version', '-i'] + list(old_packages))
+    return
+
+
+def install_packages(packages):
+    """
+    Install debian packages from dictionary
+    Input:
+        list new_packages - new packages to be installed
+    """
+    install_command = ['dpkg', '--skip-same-version', '-i'] + packages
+    subprocess.run(install_command, check=True)
+    return
+
+
+def add_path(path, packages):
+    """
+    Convert dictionary of packages into a list of paths to the debian files
+    Input:
+        string path - path to where the debian files are stored
+        dict packages - dictionary of package names
+    """
+    return [f'{path}/{package}.deb' for package in packages.keys()]
+
+
+def add_rollback_message(error):
+    """
+    Adds a message to the end of an error that notifies the user that the installation has been rolled back.
+    """
+    error.args = (error.args[0] + '\n\n    Installation has been rolled back',) + error.args[1:]
+    return error
+
+
+def make_directories(*directories):
+    for folder in directories:
+        os.makedirs(folder, exist_ok=True)
+    return
+
+
+def success_message(old_packages, new_packages):
+    return '\n\nAll packages installed successfully. Changes:\n    ' + '\n    '.join(
+        f"{package}: {old_packages.get(package,'None')} -> {version}"
+        for package, version in new_packages.items()) 
+
+
+def main(HISTORY_FOLDER, OLD_DEB_PATH, NEW_DEB_PATH):
+    make_directories(HISTORY_FOLDER, NEW_DEB_PATH, OLD_DEB_PATH)
+
+    # Extract archive, get list of packages to be installed
+    _packages = unpack_zip(os.path.dirname(__file__), unpack_path=NEW_DEB_PATH)
+
+    #              {Package_name: Version}
+    NEW_PACKAGES = {os.path.splitext(item)[0]: get_version(f'{NEW_DEB_PATH}/{item}', deb_file=True)
+                    for item in _packages}
+    OLD_PACKAGES = {package: get_version(package)
+                    for package in NEW_PACKAGES.keys()
+                    if get_version(package) is not None}
+
+    repack(OLD_PACKAGES, OLD_DEB_PATH)
```

Comment:
I think we should write more functions that returns useful outputs, rather than reprocessing old inputs.

For example, consider the following code:
```py
OLD_PACKAGES = {package: get_version(package)
                for package in NEW_PACKAGES.keys()
                if get_version(package) is not None}

repack(OLD_PACKAGES, OLD_DEB_PATH)
rollback(NEW_PACKAGES.keys(), add_path(OLD_DEB_PATH, OLD_PACKAGES))
```
`repack` can return actual paths to old versioned debs to be used by `rollback`

```py
old_deb_paths = repack(OLD_PACKAGES, OLD_DEB_PATH)
rollback(NEW_PACKAGES.keys(), old_deb_paths)
```
I think that provides smoother reading from top to bottom. Without needing to revisit old variables `OLD_PACKAGES` again to see how `add_path` transforms it.